### PR TITLE
[FIX] sale_project: prevent premature task so confirmation

### PR DESCRIPTION
**Pull request by Yoann-bary**
Steps to reproduce:
- Sales app > Configuration > Settings > Enable 'Quotation Templates'
- Default template > Search more > New
- Set 'Confirmation Mail' to 'Sales: Order Confirmation'
- Project app > New > Enable 'Billable' in the project's settings
- New Task > Set any Customer
- Sales Order Item > Search More > New > Add product then save
- Click 'Sales Order' button

Coupled with https://github.com/odoo/enterprise/pull/73492

You'll see in the chatter that the confirmation email is sent before the product is added to the sales order, resulting in a $0 quotation document being sent to the customer. This happens because we confirm the SO at the same time as we create it (After pressing the New button). Instead we would rather confirm the SO on save, after the order line creation.

The SO confirmation is necessary because we don't want people to work on tasks/projects that are still in quotation since they could hypothetically still be cancelled.

opw-4190402

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by nle-odoo — APPROVED** [review]
pre-review: could be good to me
functionally: no idea and no idea about this importance of the drawbacks and also changing the behavior in stable. But logically if we should not confirm a sale.order without lines (as this PR suggest), it would not make sense to do an `action_confirm` in a default_get (where the line doesn't yet exist), but I guess it was done for a reason.

> Unfortunately we can't really confirm the order there as we can't
guarantee it was created through project and therefore run the risk of
confirming an order twice 

There would be a way, if currently there is none there could be a context key added. But functionally I don't know what we want and why the sale.order was confirmed in a default_get.

**Comment by Yoann-bary** [comment]
I'm not really comfortable with the idea of leaving something in the context if it's not targeted specifically at the SO which is why I didn't go for that. Also I learned that confirming an SO with no lines is valid, though I have no idea why we'd want to do that.

File: addons/sale_project/models/project.py [hunk 1]
```diff
@@ -11,6 +11,16 @@
...
 
 
+def _confirm_newly_created_order(record, sale_line_id):
+    """ Orders created from project/task are supposed to be confirmed to match the typical flow from sales, but since
+    we allow SO creation from the project/task itself we want to confirm newly created SOs immediately after creation.
+    However this would leads to SOs being confirmed without a single product, so we'd rather do it on record save.
+    """
+    sale_order = record.env['sale.order.line'].browse(sale_line_id).order_id
+    if sale_order and sale_order.state == 'draft':
+        sale_order.action_confirm()
```
Comment (xavierbol) [inline addons/sale_project/models/project.py:21]: I would prefer having that function as a method in each model to be able to easily override that in a custom module (if needed)

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I left one comment. 🙂 

(My comment in your enterprise PR is also valid here) 🙂

File: addons/sale_project/models/project.py [hunk 2]
```diff
@@ -139,6 +139,35 @@ def _compute_display_sales_stat_buttons(self):
...
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        projects = super().create(vals_list)
+        sol_ids = [
+            vals['sale_line_id']
+            for vals in vals_list
+            if vals.get('sale_line_id')
+        ]
+        projects._ensure_sale_order_linked(sol_ids)
```
Comment (xavierbol) [inline addons/sale_project/models/project.py:162]:
```suggestion
        if sol_ids:
            projects._ensure_sale_order_linked(sol_ids)
```

to only call the method once it is needed. 🙂

File: addons/sale_project/models/project.py [hunk 3]
```diff
@@ -929,6 +958,35 @@ def _check_sale_line_type(self):
...
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        tasks = super().create(vals_list)
+        sol_ids = [
+            vals['sale_line_id']
+            for vals in vals_list
+            if vals.get('sale_line_id')
+        ]
+        tasks._ensure_sale_order_linked(sol_ids)
```
Comment (xavierbol) [inline addons/sale_project/models/project.py:981]: same here

```suggestion
        if sol_ids:
            tasks._ensure_sale_order_linked(sol_ids)
```

File: addons/sale_project/models/project.py [hunk 4]
```diff
@@ -139,6 +139,35 @@ def _compute_display_sales_stat_buttons(self):
...
+            quotations.action_confirm()
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        projects = super().create(vals_list)
+        sol_ids = [
+            vals['sale_line_id']
+            for vals in vals_list
+            if vals.get('sale_line_id')
+        ]
```
Comment (xavierbol) [inline addons/sale_project/models/project.py:161]: Also, the set is important here to be sure we will not have the same SOL in the list/set. 🙂 

```suggestion
        sol_ids = {
            vals['sale_line_id']
            for vals in vals_list
            if vals.get('sale_line_id')
        }
```

File: addons/sale_project/models/project.py [hunk 5]
```diff
@@ -929,6 +958,35 @@ def _check_sale_line_type(self):
...
+            quotations.action_confirm()
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        tasks = super().create(vals_list)
+        sol_ids = [
+            vals['sale_line_id']
+            for vals in vals_list
+            if vals.get('sale_line_id')
+        ]
```
Comment (xavierbol) [inline addons/sale_project/models/project.py:980]: Same here

```suggestion
        sol_ids = {
            vals['sale_line_id']
            for vals in vals_list
            if vals.get('sale_line_id')
        }
```

**Review by xavierbol — COMMENTED** [review]
LGTM, thanks for your work! 🙂 

robodoo r+

**Pull request closed without merging**